### ISSUE TITLE
Jax with diff_method "parameter-shift", "finite-diff" and non-trainable parameters

### ISSUE THREAD

**Issue by rmoyard** [PennyLaneAI/pennylane#1412]
`####` Issue description

When using Jax with diff_method "finite-diff" or "parameter-shift" with a non-trainable parameter in the circuit e.g. `QubitStateVector `(see below), Python crashes. It is probably due to the fact that you cannot specify `requires_grad` attribute to an array in Jax numpy e.g. `init_state = jnp.array([psi_0, psi_1, psi_2, psi_3])` like you would do it in pennylane numpy `init_state = np.array([psi_0, psi_1, psi_2, psi_3], requires_grad=False)`. In order to solve this problem it probably requires to make changes in the Jax interface with Pennylane. I cannot figure out why it is working fine for "adjoint", "backprop" but not for "finite-diff" and "parameter-shift". If the line`qml.QubitStateVector(init_state, wires=[0, 1])` is removed from the code below, Python is not crashing.
```   
jax = pytest.importorskip("jax")
jnp = pytest.importorskip("jax.numpy")

dev = qml.device("default.qubit", wires=2)

psi_0 = 0.1
psi_1 = 0.2
psi_2 = 0.3
psi_3 = 0.4

init_state = jnp.array([psi_0, psi_1, psi_2, psi_3])
norm = jnp.linalg.norm(init_state)
init_state = init_state / norm

@qml.qnode(dev, diff_method="parameter-shift", interface="jax")
def circuit(phi):
    qml.QubitStateVector(init_state, wires=[0, 1])
    qml.IsingXX(phi, wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

phi = jnp.array(0.1)

expected = (0.5 * (1 / norm ** 2) * (-np.sin(phi) * (psi_0 ** 2 + psi_1 ** 2 - psi_2 ** 2 - psi_3 ** 2)
        + 2 * np.sin(phi / 2) * np.cos(phi / 2) * (-(psi_0 ** 2) - psi_1 ** 2 + psi_2 ** 2 + psi_3 ** 2)))

res = jax.grad(circuit, argnums=0)(phi)
print(np.allclose(res, expected, atol=1e-3, rtol=0))
```

* *Expected behavior:* 
The code above should run fine for every different diff-method but it is not the case.

* *Actual behavior:* 
I encounter `Fatal Python error: Aborted` when Jax uses diffmethod "finite-diff" or "parameter-shift". It runs correctly when using diff-method "adjoint" or "backprop".

* *Reproduces how often:* (What percentage of the time does it reproduce?)
Every time.

* *System information:*
Platform info:           macOS-10.15.7-x86_64-i386-64bit
Python version:          3.9.5
Numpy version:           1.19.5
Scipy version:           1.6.3
  Installed devices:
  - default.gaussian (PennyLane-0.16.0.dev0)
  - default.mixed (PennyLane-0.16.0.dev0)
  - default.qubit (PennyLane-0.16.0.dev0)
  - default.qubit.autograd (PennyLane-0.16.0.dev0)
  - default.qubit.jax (PennyLane-0.16.0.dev0)
  - default.qubit.tf (PennyLane-0.16.0.dev0)
  - default.tensor (PennyLane-0.16.0.dev0)
  - default.tensor.tf (PennyLane-0.16.0.dev0)

**Comment by josh146** [comment]
Thanks for catching this @rmoyard!

> I cannot figure out why it is working fine for "adjoint", "backprop" but not for "finite-diff" and "parameter-shift". 

For `finite-diff` and `parameter-shift`, the JAX interface gets applied. This performs the following logic:

1. Input JAX `DeviceArrays` are unwrapped into NumPy arrays, and are marked as trainable
2. The forward and backward pass are computed using NumPy arrays
3. The output is converted back into `DeviceArrays`

For `backprop`, the interface _is not applied_. Instead, the entire computation is pure JAX (all the way through the device), and so no unwrapping/custom gradients should be required.

However, I'm not sure why this doesn't seem to cause an issue with `adjoint`.

**Comment by glassnotes** [comment]
I've just run into this in the development of #1552. Once it is resolved, it would be nice to update the test 
`test_gradient_unitary_to_rot_two_qubit_jax` in the `test_unitary_to_rot.py` file to include `parameter-shift` in addition to `backprop`.

**Comment by antalszava** [comment]
The original code runs well both with and without jitting. The devices checked were `default.qubit` and `lightning.qubit`. #3183 addresses the comment on `test_gradient_unitary_to_rot_two_qubit_jax`.
